### PR TITLE
Add parallelization in get_valid_actions

### PR DESCRIPTION
**Pull request by MarcCote**
Do not merge. I made this PR so we can discuss it. There's still some cleaning/refactoring that needs to be done.

### CODE REVIEW

**Comment by MarcCote** [comment]
I'm using the following script to benchmark:

```python
import ctypes
import time
import sys
import jericho
import glob
from tqdm import tqdm


def run(rom, optimize=True, parallel=False):
    env = jericho.FrotzEnv(rom)
    walkthrough = env.get_walkthrough()
    start = time.time()
    obs = env.reset()
    valid_per_step = []
    for act in tqdm(walkthrough[:25]):
        valid_acts = env.get_valid_actions(use_ctypes=optimize, use_parallel=parallel)
        valid_per_step.append(valid_acts)
        obs, rew, done, info = env.step(act)
    elapsed = time.time() - start
    env.close()
    return elapsed, valid_per_step


def compare(rom):
    t1, gt_valids = run(rom, optimize=False)
    t2, pred_valids = run(rom, optimize=True)
    t3, pred_valids_para = run(rom, optimize=False, parallel=True)
    t4, pred_valids_para_c = run(rom, optimize=True, parallel=True)
    assert len(gt_valids) == len(pred_valids)
    # assert len(gt_valids) == len(pred_valids_para)
    # assert len(gt_valids) == len(pred_valids_para_c)
    for idx, (v1, v2) in enumerate(zip(gt_valids, pred_valids)):
        false_negs = set(v1) - set(v2)
        false_pos = set(v2) - set(v1)
        for a in false_negs:
            print(f"Act {a} was valid in Python but not in CTypes")
        for a in false_pos:
            print(f"Act {a} was valid in Ctypes but not in Python")
        # if v1 != v2:
        #     print(f"Erroneous valid actions in rom {rom} at step {idx}: GT {v1} Pred {v2}")
    for idx, (v1, v2) in enumerate(zip(gt_valids, pred_valids_para)):
        false_negs = set(v1) - set(v2)
        false_pos = set(v2) - set(v1)
        for a in false_negs:
            print(f"Act {a} was valid in Python but not in Para")
        for a in false_pos:
            print(f"Act {a} was valid in Para but not in Python")
        # if v1 != v2:
        #     print(f"Erroneous valid actions in rom {rom} at step {idx}: GT {v1} Pred {v2}")
    for idx, (v1, v2) in enumerate(zip(gt_valids, pred_valids_para_c)):
        false_negs = set(v1) - set(v2)
        false_pos = set(v2) - set(v1)
        for a in false_negs:
            print(f"Act {a} was valid in Python but not in ParaC")
        for a in false_pos:
            print(f"Act {a} was valid in ParaC but not in Python")
        # if v1 != v2:
        #     print(f"Erroneous valid actions in rom {rom} at step {idx}: GT {v1} Pred {v2}")

    speedup = 100 * (t1 / t2 - 1)
    print(f"Rom {rom} Python {t1:.1f} Ctypes {t2:.1f} Speedup {speedup:.1f}%")
    speedup = 100 * (t1 / t3 - 1)
    print(f"Rom {rom} Python {t1:.1f} Para {t3:.1f} Speedup {speedup:.1f}%")
    speedup = 100 * (t1 / t4 - 1)
    print(f"Rom {rom} Python {t1:.1f} ParaC {t4:.1f} Speedup {speedup:.1f}%")


def check_correctness(rom):
    env = jericho.FrotzEnv(rom)
    walkthrough = env.get_walkthrough()
    start = time.time()
    obs = env.reset()
    for idx, act in tqdm(enumerate(walkthrough[:25])):
        valid_acts = env.get_valid_actions(use_ctypes=False)
        valid_acts_c = env.get_valid_actions(use_ctypes=True)
        valid_acts_para = env.get_valid_actions(use_ctypes=True, use_parallel=True)
        if valid_acts != valid_acts_c:
            false_negs = set(valid_acts) - set(valid_acts_c)
            for a in false_negs:
                print(f"Step {idx} Act {a} was valid in Python but not in CTypes")
            false_pos = set(valid_acts_c) - set(valid_acts)
            for a in false_pos:
                print(f"Step {idx} Act {a} was valid in Ctypes but not in Python")
                state = env.get_state()
                print(env.step(a))
                env.set_state(state)
        if valid_acts != valid_acts_para:
            false_negs = set(valid_acts) - set(valid_acts_para)
            for a in false_negs:
                print(f"Step {idx} Act {a} was valid in Python but not in Para")
            false_pos = set(valid_acts_para) - set(valid_acts)
            for a in false_pos:
                print(f"Step {idx} Act {a} was valid in Para but not in Python")
                state = env.get_state()
                print(env.step(a))
                env.set_state(state)

        obs, rew, done, info = env.step(act)

    elapsed = time.time() - start
    env.close()
    return elapsed


if __name__ == "__main__":
    # check_correctness(sys.argv[1])
    compare(sys.argv[1])

```

File: frotz/src/interface/frotz_interface.c [hunk 1]
```diff
@@ -1704,18 +1704,17 @@ void test() {
 // Given a list of action candidates, find the ones that lead to valid world changes.
 // candidate_actions contains a string with all the candidate actions, seperated by ';'
 // valid_actions will be written with each of the identified valid actions seperated by ';'
-// equiv_classes will be written with the world_diff index for each valid_action indicating
+// diff_array will be written with the world_diff for each valid_action indicating
 // which of the valid actions are equivalent to each other in terms of their world diffs.
 // Returns the number of valid actions found.
-int filter_candidate_actions(char *candidate_actions, char *valid_actions, unsigned char *equiv_classes) {
```
Comment (MarcCote) [inline frotz/src/interface/frotz_interface.c:1710]: The main difference with your C function is it keeps the world_diff for each valid action. I merge them after in the Python code. The main reason being when I do the parallelization, I split the candidate actions in `os.cpucount()` chunks which are passed to the C function. That means I can't do the full equiv_classes check as you did.

File: frotz/src/interface/frotz_interface.c [hunk 2]
```diff
@@ -1778,34 +1769,14 @@ int filter_candidate_actions(char *candidate_actions, char *valid_actions, unsig
...
-          diff_array_idx++;
-          if (diff_array_idx >= diff_arr_size) {
-            printf("Too many types of valid-actions detected. Exiting!");
-            break;
-          }
-        }
-        equiv_classes[valid_cnt] = (unsigned char) match_idx;
+
+        // Write the world diff resulting from the last action.
+        get_cleaned_world_diff(&diff_array[128*valid_cnt], &diff_array[(128*valid_cnt) + 64]);
```
Comment (MarcCote) [inline frotz/src/interface/frotz_interface.c:1779]: The world_diff is written directly into the numpy array's memory.

File: jericho/jericho.py [hunk 3]
```diff
@@ -34,6 +34,57 @@
...
+        yield fct(itertools.chain((first,), rest))
+        # Make sure rest is exhausted
+        for _ in rest:
+            pass
+
+def init_worker(story, seed):
+    worker.env = FrotzEnv(story, seed)
+
+
+def worker(data):
```
Comment (MarcCote) [inline jericho/jericho.py:61]: Just for testing. I don't think there's any reason to keep this non-C parallelized version.

File: frotz/src/interface/frotz_interface.c [hunk 4]
```diff
@@ -1778,34 +1769,14 @@ int filter_candidate_actions(char *candidate_actions, char *valid_actions, unsig
 
     if (get_score() != orig_score || game_over() > 0 || victory() > 0 || world_changed() > 0) {
```
[the remark this answers is not in the document]
Comment (mhauskn) [reply frotz/src/interface/frotz_interface.c:1770]: Yes, I think combining them could result in some time savings.

File: frotz/src/interface/frotz_interface.c (hunk 2 above)
Comment (mhauskn) [reply frotz/src/interface/frotz_interface.c:1779]: Looks good

**Comment by mhauskn** [comment]
Hey Marc, thanks for this PR - I think we should definitely include it, after any necessary cleanup/refactoring. I've double checked the benchmark scripts and am getting similar speedups on my personal machine.

**Comment by MarcCote** [comment]
Great. Let me clean it and I'll ping you later.

File: frotz/src/interface/frotz_interface.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarcCote) [reply frotz/src/interface/frotz_interface.c:1770]: I haven't done this.

File: jericho/jericho.py [hunk 5]
```diff
@@ -889,7 +911,7 @@ def _identify_interactive_objects(self, observation='', use_object_tree=False):
         self.set_state(state)
         return desc2obj
 
-    def _filter_candidate_actions(self, candidate_actions, use_ctypes):
+    def _filter_candidate_actions(self, candidate_actions, use_ctypes=False, use_parallel=False):
```
Comment (MarcCote) [inline jericho/jericho.py:914]: @mhauskn I've refactored this function a bit to include the parallel version.

**Comment by MarcCote** [comment]
Alright, I'm happy with the refactoring for now.

**Pull request merged**